near-sdk/src/collections/tree_map.rs

```diff
@@ -0,0 +1,1300 @@
+                        }
+                    }
+                    self.update_height(at);
+                    self.enforce_balance(at)
+                }
+            },
+            None => at
+        }
+    }
+
+    // Node holding the key is not removed from the tree - instead the substitute node is found,
+    // the key is copied to 'removed' node from substitute node, and then substitute node gets
+    // removed from the tree.
+    //
+    // The substitute node is either:
+    // - right-most (max) node of the left subtree (containing smaller keys) of node holding `key`
+    // - or left-most (min) node of the right subtree (containing larger keys) of node holding `key`
+    //
+    fn do_remove(&mut self, key: &K) -> u64 {
+        let root = self.root;
+        // r_id - id of a node containing key of interest
+        // r_p - id of an immediate parent node of r_id
+        let (r_id, r_p) = match self.lookup_at(root, key) {
+            Some(x) => x,
+            None => return root // cannot remove a missing key, no changes to the tree needed
+        };
+
+        let lft_opt = self.lft.get(&r_id);
+        let rgt_opt = self.rgt.get(&r_id);
+
+        if lft_opt.is_none() && rgt_opt.is_none() {
+            // remove leaf
+            let p_key = self.key.get(&r_p).unwrap();
+            if p_key.lt(key) {
+                self.rgt.remove(&r_p);
+            } else {
+                self.lft.remove(&r_p);
+            }
+            self.key.remove(&r_id);
+            self.ht.remove(&r_id);
+
+            // removing node might have caused a imbalance - balance the tree up to the root,
+            // starting from lowest affected key - the parent of a leaf node in this case
+            self.check_balance(root, &p_key)
+
+        } else {
+            // non-leaf node, select subtree to proceed with
+            let b = self.get_balance(r_id);
+            if (b >= 0 && lft_opt.is_some()) || rgt_opt.is_none() {
+                // proceed with left subtree
+                let lft = lft_opt.unwrap();
+
+                // k - max key from left subtree
+                let k = self.max_at(lft).unwrap();
+
+                // n - id of a node that holds key k, p - id of immediate parent of n
+                let (n, p) = self.lookup_at(r_id, &k).unwrap();
+
+                self.key.insert(&r_id, &k);
+                self.key.remove(&n);
+                self.ht.remove(&n);
+
+                if self.rgt.get(&p).map(|id| id == n).unwrap_or_default() {
+                    // n is on right link of p
+                    match self.lft.get(&n) {
+                        Some(l) => {
+                            self.rgt.insert(&p, &l);
+                            self.lft.remove(&n);
+                        },
+                        None => {
+                            self.rgt.remove(&p);
+                        }
+                    };
+                } else {
+                    // n is on left link of p
```

Comment:
`n` corresponds to the largest element in the subtree of `lft`. For it to be on the left link of its parent, it's parent can't be in the subtree of `lft` (otherwise it's parent is larger than `n` but `n` is the largest). Thus, the parent of `n` is `r_id` itself. Thus, below it must be that `p == r_id`. As such, the `self.lft.insert(p, l)` below is a no-op, and should be removed.